arangod/RocksDBEngine/RocksDBPersistedLog.h

```diff
@@ -34,21 +35,86 @@
 
 namespace arangodb {
 
-struct RocksDBLogPersistor : std::enable_shared_from_this<RocksDBLogPersistor> {
-  struct Executor {
-    virtual ~Executor() = default;
-    virtual void operator()(fu2::unique_function<void() noexcept>) = 0;
+struct AsyncLogWriteContext {
+  AsyncLogWriteContext(std::uint64_t vocbaseId, std::uint64_t objectId)
+      : vocbaseId(vocbaseId), objectId(objectId) {}
+  std::uint64_t const vocbaseId;
+  std::uint64_t const objectId;
+
+  void addPendingAsyncOperation();
+  void finishPendingAsyncOperation();
+  void waitForCompletion();
+
+ private:
+  std::atomic<std::size_t> _pendingAsyncOperations{0};
```

Comment:
Better to don't use std::atomic::wait/notify it contains a lot of bugs right now (deadlocked in some cases)!

Also if you want to use syscall futex, ulock, waitonaddress, it should be int!
For other types it will be use small (16 in libstdc++) static table of int for wait/wake

Reply:
Where did you get that atomic wait/notify implementations are buggy? I know that the glibc sempahore implementation is buggy, but only because the atomic wait/notify _usage_ is incorrect.

Reply:
https://gcc.gnu.org/bugzilla/buglist.cgi?quicksearch=atomic%20wait
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106183 fixed in trunk
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=101037
maybe other

Also libstdc++ semaphore don't use atomic it use separate call with platform wait/wake (because other methods have spin wait which implemented in semaphore)

Reply:
If I read that correctly, there are bugs in `libstdc++`'s fallback implementation, when `_GLIBCXX_HAVE_PLATFORM_WAIT` is false. CMIIW, but I think we're not affected; as this should be true for Linux, on Windows we're using Microsoft's STL, and on Mac we're using `libc++` (and maybe it also has platform wait, I don't know).

Reply:
@goedderz if I understand correctly you not fully right.
These bugs in "not platform atomic wait", in waiter pool, waiter pool used always when T from atomic<T> is not int (for libc++) and sizeof(int) for libstdc++.

Reply:
> and for std::atomic<T> types where T is different from __platform_wait_t.

I found, you probably missing this part from issue

Reply:
```cpp
#if !defined(_LIBCPP_VERSION) || defined(__linux__) || (defined(_AIX) && !defined(__64BIT__))
  using wait_t = std::int32_t;
#else
  using wait_t = std::int64_t;
#endif
```

Reply:
Thanks for clearing that up @MBkkt. We should probably find out then whether the fix (https://gcc.gnu.org/git/gitweb.cgi?p=gcc.git;h=af98cb88eb4be6a1668ddf966e975149bf8610b1) has made it into the libstdc++ version we're using yet. Or I'll just switch to another primitive for now.

Reply:
Unfortunately we definitely don't have such fix, it committed in this year, we use alpine with gcc 11.2 so..

But maybe int is enough for your purpose?

Anyway you are always can use condition_variable + mutex + atomic_size_t
Or absl::Mutex + atomic_size_t if you're caring about sizeof.

In general it will be works maybe better than waiters pool.

Reply:
> But maybe int is enough for your purpose?

Yes, I've decided to go with [your recommendation](https://github.com/arangodb/arangodb/pull/17528/files/ffd79b1ad531797fb504c401753c08c71933a3b9?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#r1050126579).